src/Features/Core/Portable/PdbSourceDocument/ImplementationAssemblyLookupService.cs

```diff
@@ -0,0 +1,196 @@
+            // will be in the corresponding "lib" folder.
+            var start = referenceDllPath.IndexOf(@"\ref\");
+            if (start == -1)
+                return false;
+
+            var pathToTry = referenceDllPath.Substring(0, start) +
+                            @"\lib\" +
+                            referenceDllPath.Substring(start + 5);
+
+            if (IOUtilities.PerformIO(() => File.Exists(pathToTry)))
+            {
+                implementationDllPath = pathToTry;
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool TryTargetingPackToSharedSdk(string referencedDllPath, [NotNullWhen(true)] out string? implementationDllPath)
+        {
+            implementationDllPath = null;
+
+            // eg. C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.5\ref\net6.0\Foo.dll
+            var parts = referencedDllPath.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+            if (parts is not [.., "packs", var packName, var packVersion, "ref", _, var dllFileName])
+                return false;
+
+            // We try to get the shared sdk name from the FrameworkList.xml file, in the data dir
+            // eg. C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.5\data\FrameworkList.xml
+            var frameworkXml = Path.Combine(referencedDllPath, "..", "..", "..", "data", "FrameworkList.xml");
+
+            string sdkName;
+            try
+            {
+                using var fr = File.OpenRead(frameworkXml);
+                using var xr = XmlReader.Create(fr);
+                xr.Read();
+                sdkName = xr.GetAttribute("FrameworkName");
+            }
+            catch
+            {
+                // This could be a file read error, or XML error, but we don't really care, as we're only trying to
+                // use a heuristic to provide better results, we don't have to be super resiliant to all things.
+                return false;
+            }
+
+            // If it exists, the implementation dll will be in the shared sdk folder for this pack
+            // eg. C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.5\Foo.dll
+            // But first we go up six levels to get to the common root. The pattern match above
+            // ensures this will be valid.
+            var basePath = Path.GetFullPath(Path.Combine(referencedDllPath, "..", "..", "..", "..", "..", ".."));
+            var dllPath = Path.Combine(basePath, "shared", sdkName, packVersion, dllFileName);
+
+            if (IOUtilities.PerformIO(() => File.Exists(dllPath)))
+            {
+                implementationDllPath = dllPath;
+                return true;
+            }
+
+            return false;
+        }
+
+        private string? GetExportedTypeAssemblyName(MetadataReader md, string typeFullMetadataName, string dllPath)
```

Comment:
>typeFullMetadataName [](http://example.com/codeflow?start=78&length=20)

Rather then passing in a full name consider passing in namespace and type name in separate parameters, so that you can compare them with et.Namespace and et.Name without allocating strings.